src/printer.js

```diff
@@ -3660,101 +3676,91 @@ function printJSXChildren(path, options, print, jsxWhitespace) {
   // using `map` instead of `each` because it provides `i`
   path.map((childPath, i) => {
     const child = childPath.getValue();
-    if (isLiteral(child) && typeof child.value === "string") {
-      const value = child.raw || child.extra.raw;
+    if (isLiteral(child)) {
+      const text = rawText(child);
 
       // Contains a non-whitespace character
-      if (/[^ \n\r\t]/.test(value)) {
-        // treat each line of text as its own entity
-        value.split(/(\r?\n\s*)/).forEach(textLine => {
-          const newlines = textLine.match(/\n/g);
-          if (newlines) {
-            children.push("");
-            children.push(hardline);
-
-            // allow one extra newline
-            if (newlines.length > 1) {
-              children.push("");
-              children.push(hardline);
-            }
-            return;
-          }
-
-          if (textLine.length === 0) {
-            return;
-          }
-
-          const beginSpace = /^[ \n\r\t]+/.test(textLine);
-          if (beginSpace) {
-            children.push("");
+      if (/[^ \n\r\t]/.test(text)) {
```

Comment:
I'll see what I can do to extract it and add an explanatory comment.

The short version is that these are the only four characters that count as white space in JSX. All the other chars the `\s` covers need to be treated as normal text.